File: codegen/output/c/Treasure.cpp
// -*-c++-*-

#include "Treasure.h"
#include "game.h"

#include "Pirate.h"

Treasure::Treasure(_Treasure* pointer)
{
    ptr = (void*) pointer;
}

int Treasure::id()
{
  return ((_Treasure*)ptr)->id;
}

int Treasure::x()
{
  return ((_Treasure*)ptr)->x;
}

int Treasure::y()
{
  return ((_Treasure*)ptr)->y;
}

_Pirate* Treasure::pirate()
{
  return ((_Treasure*)ptr)->pirate;
}




std::ostream& operator<<(std::ostream& stream,Treasure ob)
{
  stream << "id: " << ((_Treasure*)ob.ptr)->id  <<'\n';
  stream << "x: " << ((_Treasure*)ob.ptr)->x  <<'\n';
  stream << "y: " << ((_Treasure*)ob.ptr)->y  <<'\n';
  stream << "pirate: " << ((_Treasure*)ob.ptr)->pirate  <<'\n';
  return stream;
}

File: codegen/output/java/Treasure.java
import com.sun.jna.Pointer;

///A treasure
class Treasure: extends Mappable 
{
  Pointer ptr;
  int ID;
  int iteration;
  public Treasure(Pointer p)
  {
    ptr = p;
    ID = Client.INSTANCE.treasureGetId(ptr);
    iteration = BaseAI.iteration;
  }

  boolean validify()
  {
    if(iteration == BaseAI.iteration) return true;
    for(int i = 0; i < BaseAI.treasures.length; i++)
    {
      if(BaseAI.treasures[i].ID == ID)
      {
        ptr = BaseAI.treasures[i].ptr;
        iteration = BaseAI.iteration;
        return true;
      }
    }
    throw new ExistentialError();
  }
    
    //commands
    
    
    //getters
    
  ///Unique Identifier
  public int getId()
  {
    validify();
    return Client.INSTANCE.treasureGetId(ptr);
  }
  ///The X position of the top left corner of this object.  X is horizontal
  public int getX()
  {
    validify();
    return Client.INSTANCE.treasureGetX(ptr);
  }
  ///The Y position of the top left corner of this object.  Y is vertical
  public int getY()
  {
    validify();
    return Client.INSTANCE.treasureGetY(ptr);
  }
  ///The pirate carrying this treasure
  public Pointer getPirate()
  {
    validify();
    return Client.INSTANCE.treasureGetPirate(ptr);
  }

}
